### PR TITLE
MAISTRA-842 only update smcp status when condition status changes

### PR DESCRIPTION
**Pull request by rcernich**
Signed-off-by: rcernich <rcernich@redhat.com>

### CODE REVIEW

File: pkg/controller/servicemesh/controlplane/reconciler.go [hunk 1]
```diff
@@ -490,13 +502,30 @@ func (r *ControlPlaneReconciler) postReconciliationStatus(reconciliationReason v
...
 	} else {
-		reconciledCondition.Message = fmt.Sprintf("%s: error: %s", reconciliationMessage, processingErr)
-		r.Manager.GetRecorder(controllerName).Event(r.Instance, corev1.EventTypeWarning, reason, reconciliationMessage)
+		// grab the cause, as it's likely the error includes the reconciliation message
+		reconciledCondition.Message = fmt.Sprintf("%s: error: %s", reconciliationMessage, errors.Cause(processingErr))
+		r.Manager.GetRecorder(controllerName).Event(r.Instance, corev1.EventTypeWarning, reason, reconciledCondition.Message)
 	}
 	r.Status.SetCondition(reconciledCondition)
+
+	// we should only post status updates if condition status as changed
```
Comment (luksa) [inline pkg/controller/servicemesh/controlplane/reconciler.go]: as -> has

File: pkg/controller/servicemesh/controlplane/reconciler.go [hunk 2]
```diff
@@ -72,13 +72,20 @@ const (
 )
 
 func (r *ControlPlaneReconciler) Reconcile() (result reconcile.Result, err error) {
+	if r.Status.GetCondition(v1.ConditionTypeReconciled).Status != v1.ConditionStatusFalse { 
+		r.initializeReconcileStatus()
+		r.PostStatus()
```
Comment (luksa) [inline pkg/controller/servicemesh/controlplane/reconciler.go]: unhandled error

File: pkg/controller/servicemesh/controlplane/reconciler.go [hunk 3]
```diff
@@ -490,13 +502,30 @@ func (r *ControlPlaneReconciler) postReconciliationStatus(reconciliationReason v
...
+	if r.skipStatusUpdate() {
+		return nil
+	}
+
 	return r.PostStatus()
 }
 
+func (r *ControlPlaneReconciler) skipStatusUpdate() bool {
+	for _, conditionType := range []v1.ConditionType{v1.ConditionTypeInstalled, v1.ConditionTypeReconciled, v1.ConditionTypeReady} {
+		// TODO: what if the status is the same, but the reason is different? - we should also compare the reason
```
Comment (luksa) [inline pkg/controller/servicemesh/controlplane/reconciler.go]: I think only checking the status is correct. The reason only applies when the status actually transitions from one state to the other. We should never update the status if the state doesn't change, regardless if the underlying reason changes.

**Review by luksa — CHANGES_REQUESTED** [review]

File: pkg/controller/servicemesh/controlplane/reconciler.go (hunk 3 above)
Comment (rcernich) [reply pkg/controller/servicemesh/controlplane/reconciler.go]: removed

File: pkg/controller/servicemesh/controlplane/reconciler.go (hunk 2 above)
Comment (rcernich) [reply pkg/controller/servicemesh/controlplane/reconciler.go]: returning the error now

File: pkg/controller/servicemesh/controlplane/reconciler.go (hunk 1 above)
Comment (rcernich) [reply pkg/controller/servicemesh/controlplane/reconciler.go]: updated

**Review by luksa — APPROVED** [review]
LGTM. Events and conditions look perfect now.

**Review by knrc — APPROVED** [review]

**Pull request merged**